src/exoscale/vinyl/store.clj

```diff
@@ -302,3 +384,52 @@
    all records for a specific composite key prefix"
   [txn-context record-type items]
   (delete-by-range txn-context (all-of-range txn-context record-type items)))
+
+(defn long-query-reducer
+  "A reducer over large ranges. Accepts queries as per `execute-query`. Results
+   are reduced into an accumulator with the help of the reducing function `f`.
+   The accumulator is initiated to `init`. `clojure.core.reduced` is honored.
+
+   Obviously, this approach does away with any consistency guarantees usually
+   offered by FDB.
+
+   Results being accumulated in memory, this also means that care must be
+   taken with the accumulator."
+  ;; Some cliff notes to read the code below.
+  ;; The basic idea is that scanning is done while it works, up until
+  ;; the point where an exception will be raised.
+  ;;
+  ;; By default, `run-async` uses a simple exponential back-off algorithm
+  ;; between transaction function retries. In this case we want to avoid that.
+  ;; To that effect, a runner is created with specific parameters (a large
+  ;; `max-attempts` value, as well as minimum viable delays.
+  ;;
+  ;; We then call `apply-reduce` on the returned cursor from our query with a
+  ;; twist: every visited element will get its continuation stored in an atom.
+  ;; When interrupted, the function will be retried, which pops the last seen
+  ;; continuation.
+  ;;
+  ([db f init query]
+   (long-query-reducer db f init query {}))
+  ([db f init query {::keys [values] :as opts}]
+   (let [cont    (atom nil)
+         result  (atom init)
+         props   (execute-properties (dissoc opts ::limit))
+         ctx     (if (some? values) (query/bindings values) query/empty-context)
+         q       (as-query query)
+         runner  (wrapped-runner db
+                                 {::max-attempts        Integer/MAX_VALUE
+                                   ::max-delay           2
+                                  ::initial-delay       2})]
+     (-> (run-async
+          runner
+          (fn [^FDBRecordStore store]
+            (-> (.execute (.planQuery store q)
+                          store
+                          ctx
+                          ^bytes @cont
+                          props)
+                (cursor/apply-reduce f result #(reset! cont %)))))
```

Comment:
__Info__

This is where the magic happens to put the continuation into the atom to be able to restart the transaction at the last reading key.

Reply:
The atom is just for "convenience" right? Due to the dependency between result pages (needs the cont token for the next page) the code will run effectively serial, even if it's scheduled on some thread right?

Reply:
Yes, we may use volatile! instead.

Reply:
I'm fine with atom, I was just unsure.